splunklogger.js

```diff
@@ -369,34 +528,55 @@ SplunkLogger.prototype.use = function(middleware) {
 SplunkLogger.prototype._sendEvents = function(context, callback) {
     callback = callback || /* istanbul ignore next*/ function(){};
 
-    // Validate the context again, right before using it
-    context = this._initializeContext(context);
-    context.requestOptions.headers["Authorization"] = "Splunk " + context.config.token;
-
-    if (context.config.autoFlush) {
-        context.requestOptions.body = this._makeBody(context);
-    }
-    else {
-        // Don't run _makeBody since we've already done that
-        context.requestOptions.body = context.message;
-        // Manually set the content-type header for batched requests, default is application/json
-        // since json is set to true.
-        context.requestOptions.headers["content-type"] = "application/x-www-form-urlencoded";
-    }
     var that = this;
-    request.post(context.requestOptions, function(err, resp, body) {
-        // Call error() if error, or body isn't success
-        var error = err;
-        // Assume this is a non-success response from Splunk, build the error accordingly
-        if (!err && body && body.code.toString() !== "0") {
-            error = new Error(body.text);
-            error.code = body.code;
-        }
-        if (error) {
-            that.error(error, context);
+
+    var splunkError = null; // Errors returned by Splunk
+    var requestError = null; // Any non-Splunk errors
+
+    // References so we don't have to deal with callback parameters
+    var _response = null;
+    var _body = null;
+
+    var numRetries = 0;
+
+    utils.whilst(
+        function() {
+            // Continue if we can (re)try
+            return numRetries++ <= context.config.maxRetries;
+        },
+        function(done) {
+            that._post(context.requestOptions, function(err, resp, body) {
+                // Store the latest error, response & body
+                splunkError = null;
+                requestError = err;
+                _response = resp;
+                _body = body;
+
+                // Try to parse an error response from Splunk
+                if (!requestError && body && body.code.toString() !== "0") {
+                    splunkError = new Error(body.text);
+                    splunkError.code = body.code;
+                }
+
+                // Request error (non-200), retry if numRetries hasn't exceeded the limit
```

Comment:
There is a subtle problem here. There is no point to retry when the server returns http input error. For example if parser fails it will fail in each retry attempt.
.NET/Java have "error whitelist" logic. It's really simple ...